File: NET/Libraries/Treehopper.Libraries/Sensors/Optical/Tsl2591.cs
using System;
using System.Threading.Tasks;

namespace Treehopper.Libraries.Sensors.Optical
{
    /// <summary>
    /// Avago/ams TSL2591 High-dynamic range digital light sensor
    /// </summary>
    [Supports("ams (Avago)", "TSL2591")]
    public partial class Tsl2591 : AmbientLightSensor
    {
        /// <summary>
        /// Probes the specified I2C bus to discover any TSL2591 sensors attached.
        /// </summary>
        /// <param name="i2c">The bus to probe.</param>
        /// <param name="rate">The rate to use.</param>
        /// <returns>A TSL2591 if found, or null if not found.</returns>
        public static async Task<Tsl2591> Probe(I2C i2c, int rate=100)
        {
            var dev = new SMBusDevice(0x29, i2c, rate);
            var result = await dev.ReadByteData(0xB2);
            if (result == 0x50)
                return new Tsl2591(i2c, rate);

            return null;
        }

        private Tsl2591Registers registers;

        /// <summary>
        ///     Construct a TSL2591 ambient light sensor
        /// </summary>
        /// <param name="i2c">The bus to use.</param>
        /// <param name="rate">The rate to use.</param>
        public Tsl2591(I2C i2c, int rate=100)
        {
            registers = new Tsl2591Registers(new SMBusDevice(0x29, i2c, rate));
            registers.enable.powerOn = 1;
            registers.enable.alsEnable = 1;
            Task.Run(registers.enable.write).Wait();
        }

        public int IntegrationTimeValue
        {
            get
            {
                return (registers.config.alsTime + 1) * 100;
            }
        }

        public int GainSettingValue
        {
            get
            {
                switch (registers.config.getAlsGain())
                {
                    case AlsGains.Low:
                    default:
                        return 1;
                    case AlsGains.Medium:
                        return 25;
                    case AlsGains.High:
                        return 428;
                    case AlsGains.Max:
                        return 9876;
                }
            }
        }

        /// <summary>
        ///     Get or set the integration time of this sensor
        /// </summary>
        public AlsTimes IntegrationTime
        {
            get
            {
                return registers.config.getAlsTime();
            }
            set
            {
                if (registers.config.getAlsTime() == value)
                    return;

                registers.config.setAlsTime(value);
                Task.Run(registers.config.write).Wait();
            }
        }

        /// <summary>
        ///     Get or set the gain setting
        /// </summary>
        public AlsGains GainSetting
        {
            get
            {
                return registers.config.getAlsGain();
            }
            set
            {
                if (registers.config.getAlsGain() == value)
                    return;
                registers.config.setAlsGain(value);
                Task.Run(registers.config.write).Wait();
            }
        }

        /// <summary>
        /// Requests a reading from the sensor and updates its data properties with the gathered values.
        /// </summary>
        /// <returns>An awaitable Task</returns>
        /// <remarks>
        /// Note that when #AutoUpdateWhenPropertyRead is `true` (which it is, by default), this method is implicitly 
        /// called when any sensor data property is read from --- there's no need to call this method unless you set
        /// AutoUpdateWhenPropertyRead to `false`.
        /// 
        /// Unless otherwise noted, this method updates all sensor data simultaneously, which can often lead to more efficient
        /// bus usage (as well as reducing USB chattiness).
        /// </remarks>
        public override async Task UpdateAsync()
        {
            await registers.readRange(registers.ch0, registers.ch1);
            double cpl = (IntegrationTimeValue * GainSettingValue) / 408.0;
            this._lux = (registers.ch0.value - registers.ch1.value) * (1.0 - ((double)registers.ch1.value / registers.ch0.value)) / cpl;
        }
    }
}

File: Python/treehopper/libraries/sensors/optical/Tsl2591.py
from treehopper.api import I2C
from treehopper.libraries import SMBusDevice
from treehopper.libraries.sensors.optical import AmbientLightSensor
from treehopper.libraries.sensors.optical.Tsl2591Registers import Tsl2591Registers, AlsGains, AlsTimes


class Tsl2591(AmbientLightSensor):
    @staticmethod
    def probe(i2c: I2C):
        dev = SMBusDevice(0x29, i2c, 100)
        result = dev.read_byte_data(0xB2)
        if result == 0x50:
            return Tsl2591(i2c)

        return None

    class Gain:
        Low = 1
        Medium = 25
        High = 428
        Max = 9876

    def __init__(self, i2c):
        super().__init__()
        self._registers = Tsl2591Registers(SMBusDevice(0x29, i2c, 100))
        self._registers.enable.powerOn = 1
        self._registers.enable.alsEnable = 1
        self._registers.enable.write()

    @property
    def integration_time(self):
        return (self._registers.config.alsTime + 1) * 100.0

    @integration_time.setter
    def integration_time(self, value):
        self._registers.config.alsTime = value//100 - 1
        self._registers.config.write()

    @property
    def gain(self):
        if self._registers.config.alsGain == AlsGains.Low:
            return 1
        if self._registers.config.alsGain == AlsGains.Medium:
            return 25
        if self._registers.config.alsGain == AlsGains.High:
            return 428
        if self._registers.config.alsGain == AlsGains.Max:
            return 9876

    @gain.setter
    def gain(self, value):
        if value == Tsl2591.Gain.Low:
            self._registers.config.alsGain = 0
        if value == Tsl2591.Gain.Medium:
            self._registers.config.alsGain = 1
        if value == Tsl2591.Gain.High:
            self._registers.config.alsGain = 2
        if value == Tsl2591.Gain.Max:
            self._registers.config.alsGain = 3

        self._registers.config.write()

    def update(self):
        self._registers.readRange(self._registers.ch0, self._registers.ch1)
        if self._registers.ch0.value == 0 or self._registers.ch1.value == 0:
            return

        cpl = (self.integration_time * self.gain) / 408.0
        self._lux = (self._registers.ch0.value - self._registers.ch1.value) * \
                    (1.0 - (self._registers.ch1.value / self._registers.ch0.value)) / cpl
